src/mscorlib/src/System/Diagnostics/Eventing/EventPipe.cs

```diff
@@ -0,0 +1,157 @@
+using System.Runtime.InteropServices;
+using System.Security;
+using Microsoft.Win32;
+
+namespace System.Diagnostics.Tracing
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct EventPipeProviderConfiguration
+    {
+        [MarshalAs(UnmanagedType.LPWStr)]
+        private string m_providerName;
+        private UInt64 m_keywords;
+        private uint m_loggingLevel;
+
+        internal EventPipeProviderConfiguration(
+            string providerName,
+            UInt64 keywords,
+            uint loggingLevel)
+        {
+            if(string.IsNullOrEmpty(providerName))
+            {
+                throw new ArgumentNullException(nameof(providerName));
+            }
+            if(loggingLevel > 5) // 5 == Verbose, the highest value in EventPipeLoggingLevel.
+            {
+                throw new ArgumentOutOfRangeException(nameof(loggingLevel));
+            }
+            m_providerName = providerName;
+            m_keywords = keywords;
+            m_loggingLevel = loggingLevel;
+        }
+
+        internal string ProviderName
+        {
+            get { return m_providerName; }
+        }
+
+        internal UInt64 Keywords
+        {
+            get { return m_keywords; }
+        }
+
+        internal uint LoggingLevel
+        {
+            get { return m_loggingLevel; }
+        }
+    }
+
+    internal sealed class EventPipeConfiguration
+    {
+        private string m_outputFile;
+        private uint m_circularBufferSizeInMB;
+        private List<EventPipeProviderConfiguration> m_providers;
+
+        internal EventPipeConfiguration(
+            string outputFile,
+            uint circularBufferSizeInMB)
+        {
+            if(string.IsNullOrEmpty(outputFile))
+            {
+                throw new ArgumentNullException(nameof(outputFile));
+            }
+            if(circularBufferSizeInMB == 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(circularBufferSizeInMB));
+            }
+            m_outputFile = outputFile;
+            m_circularBufferSizeInMB = circularBufferSizeInMB;
+            m_providers = new List<EventPipeProviderConfiguration>();
+        }
+
+        internal string OutputFile
+        {
+            get { return m_outputFile; }
+        }
+
+        internal uint CircularBufferSizeInMB
+        {
+            get { return m_circularBufferSizeInMB; }
+        }
+
+        internal EventPipeProviderConfiguration[] Providers
+        {
+            get { return m_providers.ToArray(); }
+        }
+
+        internal void EnableProvider(string providerName, UInt64 keywords, uint loggingLevel)
+        {
+            m_providers.Add(new EventPipeProviderConfiguration(
+                providerName,
+                keywords,
+                loggingLevel));
+        }
+    }
+
+    internal static class EventPipe
+    {
+        internal static void Enable(EventPipeConfiguration configuration)
+        {
```

Comment:
I am also conflicted about the model that is presented on enabling.  The way you have it currently you have to supply all the configuration information in one call.   This configuration is moderately complex (a list of structs that have strings inside them).   This in turn causes you to defined 'config' classes for this information and use non-trivial interop features (auto-marshaling of strings in structs).  

The alternative is in fact what ETW does.   Basically you supply configuration data as a series of calls (Enabling or disabling one provider at a time).  This is actually more flexible, and useful, and happens to also avoid the need for a 'config' structure (all the fields becomes arguments to the 'EnableProvider' call), and the non-trivial interop features (since now you call functions that take primitive or string values).   

As I said I am conflicted, but my instincts say that there was no reason to deviate from what ETW/EventListener did (having one model throughout is good).  

The more I think about it, the more I would prefer to following the ETW/EventListener pattern (you have a 'EnableProvider' calls you make).  The reason is that we will want to have EventListener be able to control EventPipes, and we don't want any unnecessary model mismatches.